### PR TITLE
BloomFilter v2 support for Spark's bloom-filter based joins

### PR DESCRIPTION
**Pull request by mythrocks**
### Description

This commit adds support for the v2 format of the BloomFilters used in Apache Spark 4.1.1 for joins (via https://github.com/apache/spark/commit/a08d8b093c0ec09cc6ce2c3642502f4842aebd86).

### Background

The v1 format used INT32s for bit index calculation. When the number of items in the bloom-filter approaches INT_MAX, one sees a higher rate of collisions. The v2 format uses INT64 values for bit index calculations, allowing the full bit space to be addressed. Apparently, this reduces the false positive rates for large filters.

Before the fix in this current PR, `spark-rapids-jni` supported only the v1 bloom filter format.  Testing `spark-rapids` on Apache Spark 4.1.1 revealed failures in mixed-mode execution, where bloom filters built on CPU were probed on the GPU (assuming v1 format).

The changes here _should_ allow for a reduced false-positive rate for bloom filters built on join keys with high cardinalities (approaching INT_MAX).  Note also that support for the v1 format is retained, for backward compatibility.

### CODE REVIEW

**Comment by mythrocks** [comment]
@greptile-apps:  Please review, to see if your prior concerns have been addressed.

File: src/main/cpp/src/bloom_filter.cu [hunk 1]
```diff
@@ -73,178 +78,246 @@ CUDF_KERNEL void gpu_bloom_filter_put(cudf::bitmask_type* const bloom_filter,
...
+        if (!(bloom_filter[word_index] & mask)) { return false; }
+      }
+    } else {
+      int64_t combined_hash =
+        static_cast<int64_t>(h1) * static_cast<int64_t>(std::numeric_limits<int32_t>::max());
+      for (int idx = 0; idx < num_hashes; idx++) {
+        combined_hash += h2;
+        int64_t combined_index        = combined_hash < 0 ? ~combined_hash : combined_hash;
+        auto const bit_pos            = combined_index % bloom_filter_bits;
+        auto const [word_index, mask] = gpu_bit_to_word_mask(bit_pos);
```
[the remark this answers is not in the document]
Comment (mythrocks) [reply src/main/cpp/src/bloom_filter.cu]: @greptile-apps:  This is also a false alarm.  

Per the Java/Spark [implementation](https://github.com/apache/spark/blob/5075ea6a85f3f1689766cf08a7d5b2ce500be1fb/common/sketch/src/main/java/org/apache/spark/util/sketch/BloomFilterImplV2.java#L63), there is no usage of `Math.abs`.  The computation is identical and compatible.

File: src/main/cpp/src/BloomFilterJni.cpp [hunk 2]
```diff
@@ -20,17 +20,26 @@
...
-  JNIEnv* env, jclass, jint numHashes, jlong bloomFilterBits)
+  JNIEnv* env, jclass, jint version, jint numHashes, jlong bloomFilterBits, jint seed)
 {
   JNI_TRY
   {
     cudf::jni::auto_set_device(env);
 
-    int bloom_filter_longs = static_cast<int>((bloomFilterBits + 63) / 64);
-    auto bloom_filter      = spark_rapids_jni::bloom_filter_create(numHashes, bloom_filter_longs);
+    jlong bloom_filter_longs_long = (bloomFilterBits + 63) / 64;
```
[the remark this answers is not in the document]
Comment (mythrocks) [reply src/main/cpp/src/BloomFilterJni.cpp]: @greptile-apps:  The use of a `jlong` here is an artifact of the previous implementation.  I would like to keep the `jlong`.
In `bloom_filter.cu`, the third parameter to `bloom_filter_create` is an `int`. So the real overflow check should be for type `int`, not `long`.  I am inclined to change this as follows:
```suggestion
    if (bloomFilterBits > static_cast<jlong>(std::numeric_limits<int>::max()) - 63) {
      JNI_THROW_NEW(env, "java/lang/IllegalArgumentException",
                    "bloom filter bit count overflows int when converted to longs", 0);
    }
    jlong bloom_filter_longs_long = (bloomFilterBits + 63) / 64;
```
This catches the overflow when adding 63.  Does this sound agreeable?

[the remark this answers is not in the document]
Comment (mythrocks) [reply src/main/cpp/src/BloomFilterJni.cpp]: @greptile-apps:  This has been resolved in the latest commit.

File: src/main/cpp/src/bloom_filter.cu [hunk 3]
```diff
@@ -73,178 +78,250 @@ CUDF_KERNEL void gpu_bloom_filter_put(cudf::bitmask_type* const bloom_filter,
...
                                                        int bloom_filter_longs,
+                                                       int seed,
                                                        rmm::cuda_stream_view stream,
                                                        rmm::device_async_resource_ref mr)
 {
-  auto [bloom_filter_size, buf_size] = get_bloom_filter_stride(bloom_filter_longs);
+  CUDF_EXPECTS(version == bloom_filter_version_1 || version == bloom_filter_version_2,
+               "Bloom filter version must be 1 or 2");
+
+  CUDF_EXPECTS(bloom_filter_longs >= 0, "Bloom filter bit count must be non-negative");
```
[the remark this answers is not in the document]
Comment (mythrocks) [reply src/main/cpp/src/bloom_filter.cu]: @greptile-apps:  Fair point.  Fixed in the latest commit.

**Comment by mythrocks** [comment]
@greptile-apps, how is this change looking now?

**Comment by mythrocks** [comment]
Build

File: src/main/cpp/benchmarks/bloom_filter.cu [hunk 4]
```diff
@@ -22,17 +22,51 @@
 #include <hash/hash.hpp>
 #include <nvbench/nvbench.cuh>
 
-static void bloom_filter_put(nvbench::state& state)
+static void bloom_filter_put_v1(nvbench::state& state)
```
Comment (ttnghia) [inline src/main/cpp/benchmarks/bloom_filter.cu]: `static` is not needed.

**Review by amahussein — COMMENTED** [review]
Thanks @ttnghia I had a couple of questions on your comments.

File: src/main/cpp/src/bloom_filter.cu [hunk 5]
```diff
@@ -256,61 +334,89 @@ void bloom_filter_put(cudf::list_scalar& bloom_filter,
...
-  auto dst = reinterpret_cast<cudf::bitmask_type*>(reinterpret_cast<uint8_t*>(buf.data()) +
-                                                   bloom_filter_header_size);
+  auto src = lcv.child().data<uint8_t>() + hdr_size;
+  auto dst =
+    reinterpret_cast<cudf::bitmask_type*>(reinterpret_cast<uint8_t*>(buf.data()) + hdr_size);
 
-  // bitwise-or all the bloom filters together
   cudf::size_type num_words = header.num_longs * 2;
   thrust::transform(
     rmm::exec_policy(stream),
```
[the remark this answers is not in the document]
Comment (ttnghia) [reply src/main/cpp/src/bloom_filter.cu]: Any pre-commit or rule would be helpful.

File: src/main/cpp/src/bloom_filter.cu [hunk 6]
```diff
@@ -72,178 +78,250 @@ CUDF_KERNEL void gpu_bloom_filter_put(cudf::bitmask_type* const bloom_filter,
...
+  } else {
+    // https://github.com/apache/spark/blob/5075ea6a85f3f1689766cf08a7d5b2ce500be1fb/common/sketch/src/main/java/org/apache/spark/util/sketch/BloomFilterImplV2.java#L63
+    int64_t combined_hash =
+      static_cast<int64_t>(h1) * static_cast<int64_t>(std::numeric_limits<int32_t>::max());
+    for (int idx = 0; idx < num_hashes; idx++) {
+      combined_hash += h2;
+      int64_t combined_index        = combined_hash < 0 ? ~combined_hash : combined_hash;
+      auto const bit_pos            = combined_index % bloom_filter_bits;
+      auto const [word_index, mask] = gpu_bit_to_word_mask(bit_pos);
+      atomicOr(bloom_filter + word_index, mask);
```
[the remark this answers is not in the document]
Comment (mythrocks) [reply src/main/cpp/src/bloom_filter.cu]: Switched to `atomic_ref.fetch_or`.  TIL that you could also use `|=` for more readability.

File: src/main/cpp/tests/bloom_filter.cu [hunk 7]
```diff
@@ -170,3 +178,175 @@ TEST_F(BloomFilterTest, ProbeMerged)
...
+  cudf::test::fixed_width_column_wrapper<int64_t> input{20, 80, 100, 99, 47, -9, 234000000};
+  spark_rapids_jni::bloom_filter_put(*bloom_filter, input, stream);
+
+  cudf::test::fixed_width_column_wrapper<int64_t> probe_in{20, 80, 100, 99, 47, -9, 234000000};
+  cudf::test::fixed_width_column_wrapper<bool> expected_in{1, 1, 1, 1, 1, 1, 1};
+  auto result = spark_rapids_jni::bloom_filter_probe(probe_in, *bloom_filter, stream);
+  CUDF_TEST_EXPECT_COLUMNS_EQUAL(expected_in, *result);
+}
+
+TEST_F(BloomFilterTest, BuildWithNullsAndProbeV2)
```
Comment (ttnghia) [inline src/main/cpp/tests/bloom_filter.cu]: We should also have test for the null cases.
Suggestion: Add a BuildAndProbeWithNullsV2 test mirroring the V1 version: build from a fully non-null input, then probe with a column containing nulls.

File: src/main/cpp/src/bloom_filter.cu [hunk 8]
```diff
@@ -72,178 +80,252 @@ CUDF_KERNEL void gpu_bloom_filter_put(cudf::bitmask_type* const bloom_filter,
...
     cudf::device_span<uint8_t>{const_cast<uint8_t*>(bloom_filter.data<uint8_t>()),
                                static_cast<size_t>(bloom_filter.size())},
     stream);
 }
 
+std::tuple<bloom_filter_header, cudf::device_span<cudf::bitmask_type>, int64_t, int32_t>
+unpack_bloom_filter(cudf::device_span<uint8_t const> bloom_filter, rmm::cuda_stream_view stream)
+{
+  return unpack_bloom_filter(
+    cudf::device_span<uint8_t>{const_cast<uint8_t*>(bloom_filter.data()), bloom_filter.size()},
```
[the remark this answers is not in the document]
Comment (mythrocks) [reply src/main/cpp/src/bloom_filter.cu]: Yeah, I think we'll need to separate the contexts here.  In the probe case, the bloom filter buffer can be const.  In the merge/put cases, these need to be writable.

File: src/main/cpp/src/bloom_filter.cu [hunk 9]
```diff
@@ -72,178 +81,247 @@ CUDF_KERNEL void gpu_bloom_filter_put(cudf::bitmask_type* const bloom_filter,
...
+                                                       int seed,
                                                        rmm::cuda_stream_view stream,
                                                        rmm::device_async_resource_ref mr)
 {
-  auto [bloom_filter_size, buf_size] = get_bloom_filter_stride(bloom_filter_longs);
+  SRJ_FUNC_RANGE();
+  CUDF_EXPECTS(version == bloom_filter_version_1 || version == bloom_filter_version_2,
+               "Bloom filter version must be 1 or 2");
+
+  CUDF_EXPECTS(bloom_filter_longs > 0, "Bloom filter bit count must be positive");
```
Comment (ttnghia) [inline src/main/cpp/src/bloom_filter.cu]: No guard against `num_hashes == 0`. A degenerate bloom filter would silently return incorrect results.
